[Bugfix][Core] Prefix caching causes incorrect outputs due to outdated ComputedBlocksTracker

quanliu1991 opened the pull request:
Prefix caching enabled causes incorrect outputs due to stale num_new_tokens_cached in ComputedBlocksTracker

**Description**

We encountered an **incorrect output issue** when **prefix caching is enabled**. After investigation, we found that there's an inconsistency between the `num_new_tokens_cached` retrieved during `schedule_prefills` and the actual tokens computed and allocated in `_allocate_and_set_running(seq_group)`.

This inconsistency leads to **wrong block assignments** in the prefix cache and subsequently **incorrect inference results**.

---

**Root Cause Analysis**

1. In the `schedule_prefills` stage, the scheduler retrieves `num_new_tokens_cached` for a sequence and records it in `ComputedBlocksTracker._seq_id_to_num_tokens_computed`.
2. However, if the sequence fails to be scheduled (e.g., due to lack of resources), the `num_tokens_computed` entry is **not cleared** from `ComputedBlocksTracker`.
3. On the next scheduling attempt, the scheduler directly reads `num_new_tokens_cached` from the **stale `_seq_id_to_num_tokens_computed`**, rather than recalculating it from the **actual prefix cached blocks**.
4. If the prefix cache blocks have **changed** during this period (e.g., from other sequences or eviction), the result becomes inconsistent and incorrect.

---

**Proposed Fix**

We submitted a PR that adds a call to `free_seq_cached_tokens()` in the following places to ensure consistency:

- Before the `break` in `_schedule_prefills` when a sequence fails to be scheduled
- In `_schedule_priority_preemption` when a waiting sequence has been queried by the 、`_get_num_new_uncached_and_cached_tokens()` method

This ensures that stale `num_new_tokens_cached` is removed, and future scheduling will correctly re-compute the value based on the current state of prefix cache blocks.

---

**Reproduction**

This issue can be reproduced under the following conditions:

- Prefix caching is enabled
- Sequences are frequently scheduled and preempted
- Multiple sequences share similar prefix tokens but scheduling failures occur

---

**Expected Behavior**

Prefix cache blocks should match the actual token count computed for the sequence in `ComputedBlocksTracker`, and no stale cached token metadata should remain if scheduling fails.

---

**Additional Notes**

This fix ensures that cached token tracking is consistent with actual prefix cache state, especially under preemption or failure conditions, thus improving the correctness of prefix cache reuse logic.

FIX  #18955 

<!--- pyml disable-next-line no-emphasis-as-heading -->

Commit:
[Bugfix][core] Prefix caching enabled causes incorrect outputs due to stale num_new_tokens_cached in ComputedBlocksTracker

Signed-off-by: 刘全 <quan.liu2@dbappsecurity.com.cn>

Commit:
[Misc]Fix typo

Signed-off-by: 刘全 <quan.liu2@dbappsecurity.com.cn>

aarnphm commented:
Csn u check if this behavior persist with c1? This is v0 codepath

zh-jp commented:
@aarnphm I follow the steps mentioned in [18955](https://github.com/vllm-project/vllm/issues/18955) and set `VLLM_USE_V1=1`, this problem does not exist.

aarnphm reviewed (approved):
I think we can stamp this for v0 bugfix, cc @DarkLight1337

heheda12345 commented:
Hi @DarkLight1337 ,
https://github.com/vllm-project/vllm/blob/af7fc84fd2e9b0fb70bc6349730b3ee73f0e1f8d/vllm/core/scheduler.py#L1143-L1150
For this part, if `if using_prompt_embeds != seq_group.uses_prompt_embeds():` happens, how will the request processed? Will the scheduler try to scheduled it again or mark it as finished?

heheda12345 commented on vllm/core/block_manager.py:
```diff
@@ -269,6 +269,10 @@ def free(self, seq: Sequence) -> None:
         self.block_tables[seq_id].free()
         del self.block_tables[seq_id]
 
+    def free_seq_cached_tokens(self, seq: Sequence) -> None:
```
```suggestion
    def remove_seq_from_computed_blocks_tracker(self, seq: Sequence) -> None:
```
free_seq_cached_tokens is a little ambiguous as people may think this function is freeing the block of some cached tokens in block_manager.
Does this function name look better for you?

heheda12345 commented on vllm/core/scheduler.py:
```diff
@@ -1687,6 +1692,20 @@ def free_seq(self, seq: Sequence) -> None:
+        seqs = seq_group.get_seqs(status=SequenceStatus.WAITING)
+        for seq in seqs:
+            self._free_seq_cached_tokens(seq)
+            seq.reset_state_for_recompute()
+        self._free_seq_group_cross_attn_blocks(seq_group)
```
why do we need these two lines?

quanliu1991 replied:
The code is not need, it had deleted.

heheda12345 reviewed (commented):
Thanks for catching this bug. Can you write some unit test for it?
And some questions:
1. Do we need to free in this branch?
https://github.com/vllm-project/vllm/blob/af7fc84fd2e9b0fb70bc6349730b3ee73f0e1f8d/vllm/core/scheduler.py#L884-L890
And also not sure about here
https://github.com/vllm-project/vllm/blob/af7fc84fd2e9b0fb70bc6349730b3ee73f0e1f8d/vllm/core/scheduler.py#L1143-L1150
2. For requests be marked as `FINISHED_IGNORED`, will ComputedBlocksTracker free it? If not freed, maybe there will be some memory leak due to the bug you report.
3. Can you help to analyze whether we need to free the tracker for other calls of `_get_num_new_uncached_and_cached_tokens` in scheduler?

DarkLight1337 commented:
> Hi @DarkLight1337 ,
> 
> https://github.com/vllm-project/vllm/blob/af7fc84fd2e9b0fb70bc6349730b3ee73f0e1f8d/vllm/core/scheduler.py#L1143-L1150
> 
> 
> For this part, if `if using_prompt_embeds != seq_group.uses_prompt_embeds():` happens, how will the request processed? Will the scheduler try to scheduled it again or mark it as finished?

From my understanding, this forces the scheduler to schedule the remaining requests in the next step. This is so that for each step, all the requests have either `using_prompt_embeds=False` or `using_prompt_embeds=True`, but not a mix of both. cc @qthequartermasterman

heheda12345 commented:
Thanks @DarkLight1337 . @quanliu1991 Then I think we may need to add free logic to this code block too.

aarnphm reviewed (changes requested):
I will wait until Chen's comments to be resolved before merging this in.

Commit:
[Bugfix][core] Prefix caching enabled causes incorrect outputs due to stale num_new_tokens_cached in ComputedBlocksTracker

Signed-off-by: 刘全 <quan.liu2@dbappsecurity.com.cn>

Commit:
add unit test for remove_seq_from_computed_blocks_tracker

Signed-off-by: 刘全 <quan.liu2@dbappsecurity.com.cn>

quanliu1991 commented:
@aarnphm @heheda12345 We have done a comprehensive check of all `_get_num_new_uncached_and_cached_tokens` call code within the schedule method, and added remove_seq calls in 5 additional branches.

1. Summary of the 9 branches where `remove_seq_from_computed_blocks_tracker()` was added:
-  _schedule_swapped:
https://github.com/vllm-project/vllm/blob/eaa2e51088d4daf36d47e566ad90e812f80e91b8/vllm/core/scheduler.py#L900-L904

- _schedule_priority_preemption:
Added remove_seq for waiting quene.
- _schedule_prefills:
Added remove_seq for 3 break branches.
Added remove_seq for 2 continue branches where leftover_waiting_sequences.appendleft(seq_group) is called.
Added remove_seq for 2 branches marked with FINISHED_IGNORED status.

2. Regarding FINISHED_IGNORED:
We did not find any place where seq_groups marked as FINISHED_IGNORED are freed.
In contrast, we did confirm that running sequences that finish sequences are freed, as are those in _async_stopped.
Therefore, we added a remove_seq call for FINISHED_IGNORED sequences.

3. On `_schedule_running` and the `No budget => Stop` case
I think the blocks used by seq_groups in the running queue won’t be evicted, so I didn’t add a remove_seq call.
But I’m still not sure when the "No budget => Stop" case would occur.
https://github.com/vllm-project/vllm/blob/eaa2e51088d4daf36d47e566ad90e812f80e91b8/vllm/core/scheduler.py#L712-L724

Commit:
[Misc]Fix typo

Signed-off-by: 刘全 <quan.liu2@dbappsecurity.com.cn>

quanliu1991 committed:
Merge branch 'vllm-project:main' into prefix-caching-enabled-causes-incorrect

heheda12345 reviewed (approved):
LGTM! Thanks for the detailed analysis.
@zh-jp @qthequartermasterman  Does this PR fix your problem?

noooop commented:
@quanliu1991 

this pr will break pooling models (embeddings models)

https://buildkite.com/vllm/ci/builds/22110#01977721-962e-413e-982f-5211bbecb3d6

quanliu1991 commented:
@noooop Sorry, PlaceholderBlockSpaceManager lacks this method. I need to consider the scope of use of this method and make reasonable modifications.
